plugins/mmdblookup/mmdblookup.c

```diff
@@ -133,6 +136,135 @@ void close_mmdb(MMDB_s *mmdb) {
 	MMDB_close(mmdb);
 }
 
+const char *SEP = "!";
+#define MAX_PATH_ELEMENTS 7
+
+// This is a thin wrapper function for MMDB_aget_value that knows how to convert
+// the rsyslog field into MMDB path, e.g., "!continent!code" -> ["continent", "code"].
+int mmdb_get_value(
+	MMDB_entry_s *const start,
+	MMDB_entry_data_s *const entry_data,
+	char *field
+) {
+	const char **path = calloc(MAX_PATH_ELEMENTS + 1, sizeof(const char *));
+	if (NULL == path) {
+		return MMDB_OUT_OF_MEMORY_ERROR;
+	}
+
+	// Split the field into path elements to construct a path array.
+	int i = 0;
+	char *brkf;
+	char *path_elem = strtok_r(field, SEP, &brkf);
+	while (NULL != path_elem && i < MAX_PATH_ELEMENTS) {
+		path[i] = path_elem;
+		i++;
+		path_elem = strtok_r(NULL, SEP, &brkf);
+	}
+	path[i] = NULL;
+
+	int status = MMDB_aget_value(start, entry_data, path);
+
+	free(path);
+
+	return status;
+}
+
+// Helper function to convert MMDB_entry_data_s to json_object.
+struct json_object *mmdb_entry_to_json(const MMDB_entry_data_s *entry) {
+	if (!entry || entry->offset == 0) {
+		return NULL;
+	}
+
+	switch (entry->type) {
+	case MMDB_DATA_TYPE_UTF8_STRING:
+		return json_object_new_string_len(entry->utf8_string, entry->data_size);
+	case MMDB_DATA_TYPE_BYTES:
+		char *hex_string = calloc((entry->data_size * 2) + 1, sizeof(char));
+		if (NULL == hex_string) {
+			dbgprintf("Error converting MMDB bytes to hex string\n");
+			return NULL;
+		}
+
+		for (uint32_t i = 0; i < entry->data_size; i++) {
+			sprintf(hex_string + (2 * i), "%02X", entry->bytes[i]);
+		}
+
+		struct json_object *json_string = json_object_new_string_len(
+			hex_string,
+			entry->data_size * 2
+		);
+		free(hex_string);
+
+		return json_string;
+	case MMDB_DATA_TYPE_DOUBLE:
+		return json_object_new_double(entry->double_value);
+	case MMDB_DATA_TYPE_FLOAT:
+		return json_object_new_double((double)entry->float_value);
+	case MMDB_DATA_TYPE_BOOLEAN:
+		return json_object_new_boolean(entry->boolean);
+	case MMDB_DATA_TYPE_UINT16:
+		return json_object_new_int((int)entry->uint16);
+	case MMDB_DATA_TYPE_UINT32:
+		return json_object_new_int((int)entry->uint32);
+	case MMDB_DATA_TYPE_INT32:
+		return json_object_new_int(entry->int32);
+	case MMDB_DATA_TYPE_UINT64:
+		return json_object_new_int64((int64_t)entry->uint64);
+	case MMDB_DATA_TYPE_UINT128: {
+		// json-c doesn't have a direct 128-bit integer type.
+		// Representing it as a string might be the best approach for lossless conversion.
+		char buffer[40];
+		snprintf(
+			buffer,
+			sizeof(buffer),
+			"%llu%016llu",
+			(unsigned long long)(entry->uint128 >> 64),
+			(unsigned long long)(entry->uint128 & 0xFFFFFFFFFFFFFFFFULL)
+		);
+		return json_object_new_string(buffer);
+	}
+	case MMDB_DATA_TYPE_ARRAY: {
+		json_object *json_array = json_object_new_array();
+		MMDB_entry_data_list_s *list = entry;
```

Comment:
This looks wrong to me. `entry` is a `MMDB_entry_data_s`, not a `MMDB_entry_data_list_s`. I am surprised it compiles. Are there a bunch of warnings?

Looking at the functions in `libmaxminddb`, I think you will need to use `MMDB_get_entry_data_list` rather than one of the `MMDB_*get_value` lookup functions if we need to support arrays and maps. it is probably easiest to convert the whole thing to a `json_object`. The performance would be better if you handled the path stuff before doing that, but it would still be a big improvement if you didn't.